### PR TITLE
enhancement(networking): Implement TCP keepalive `time` configuration

### PR DESCRIPTION
**Pull request by pablosichert**
Closes https://github.com/timberio/vector/issues/4816.

I tried to include an end-to-end test initially, but figured out that TCP keepalive packets with empty data are not visible from the `read`/`recv` call, since a length of `0` for the read buffer also indicates a closed connection on TCP sockets.

To test keepalive probes the IP packets would need to be inspected directly, which is not possible with the currently used abstraction of tokio streams.

This is acceptable if we rely on tokio implementing the keepalive mechanism correctly.

### CODE REVIEW

**Comment by jamtur01** [comment]
@pablosichert Looks like a failed check here: https://github.com/timberio/vector/pull/5157/checks?check_run_id=1434479691

File: docs/reference/components/sinks/socket.cue [hunk 1]
```diff
@@ -24,6 +24,12 @@ components: sinks: socket: {
 					enum: ["json", "text"]
 				}
 			}
+			keepalive: {
+				enabled:  true
+				time:     7200
+				interval: 75
+				retries:  5
+			}
```
Comment (binarylogic) [inline docs/reference/components/sinks/socket.cue:32]: @pablosichert we'll need to add this block to add relevant sources/sinks.

File: docs/reference/components/sinks.cue [hunk 2]
```diff
@@ -306,6 +306,46 @@ components: sinks: [Name=string]: {
...
+							time: {
+								common:      false
+								description: "The time a connection needs to be idle before sending keepalive probes."
+								required:    false
+								type: uint: {
+									default: sinks[Name].features.send.keepalive.time
+									unit:    "seconds"
+								}
+							}
+							interval: {
```
Comment (pablosichert) [inline docs/reference/components/sinks.cue:326]: Maybe we should omit the `interval` and `retries` docs until https://github.com/timberio/vector/issues/4957 is resolved.

Comment (binarylogic) [reply docs/reference/components/sinks.cue:326]: Ah yes, we should. I thought they were supported since I saw them in tests.

Comment (pablosichert) [reply docs/reference/components/sinks.cue:326]: Yes, I was misleading a bit there. I left a note on the `TcpKeepaliveConfig` struct.

**Comment by pablosichert** [comment]
I just removed the keepalive defaults from the docs as it doesn't reflect the current code, but on a second thought: Should we enable TCP keepalive by default instead?

From the RFC:

> A TCP keep-alive mechanism should only be invoked in server applications that might otherwise hang indefinitely and consume resources unnecessarily if a client crashes or aborts a connection during a network failure.

I first interpreted it to mean that it should only be activated on known bad connections. But it seems to qualify for our case regardless.

OTOH it says:

> If keep-alives are included, the application MUST be able to turn them on or off for each TCP connection, and they MUST default to off.

> The TCP specification does not include a keep-alive mechanism because it could: (1) cause perfectly good connections to break during transient Internet failures; (2) consume unnecessary bandwidth ("if no one is using the connection, who cares if it is still good?"); and (3) cost money for an Internet path that charges for packets.

**Comment by binarylogic** [comment]
cc @lukesteensen and @jszwedko for that one.

**Comment by fanatid** [comment]
Code looks good, but CI needs to be fixed.

File: docs/reference/components.cue [hunk 3]
```diff
@@ -810,6 +818,10 @@ components: {
 						receive_context: "Enriches data with useful \(features.receive.from.service.name) context."
 					}
 
+					if features.receive.keepalive.enabled != _|_ {
```
Comment (Hoverbear) [inline docs/reference/components.cue:821]: I am still getting used to cue 🤣

File: src/tcp.rs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+use std::time::Duration;
+
+/// Configuration for keepalive probes in a TCP stream.
+///
+/// This config's properties map to TCP keepalive properties in Tokio/Mio:
+/// https://github.com/tokio-rs/mio/blob/e6e403fe2a4fc14dfbc74dbb3ae3a14e3044eb6f/src/net/tcp/socket.rs#L25-L46
+///
+/// # Note
+///
+/// Support for the `interval` and `retries` options has just landed in Mio and they are not
```
Comment (jszwedko) [inline src/tcp.rs:11]: I might just drop these unimplemented options from the below struct to reduce confusion.

**Review by jszwedko — APPROVED** [review]
Nice work!

I think we should stick to the spec and leave them disabled by default.

As @fanatid noted, it looks like there are a few failing CI checks, probably due to dropping  `#[cfg(feature = "listenfd")]`. There is a CI task that tries to build each component independently and it seems like that is failing for at least `aws_kinesis_firehose`.

File: src/sources/socket/mod.rs [hunk 5]
```diff
@@ -34,7 +34,11 @@ pub enum Mode {
 }
 
 impl SocketConfig {
-    pub fn make_tcp_config(addr: SocketAddr) -> Self {
+    pub fn make_tcp_config(tcp_config: tcp::TcpConfig) -> Self {
```
Comment (Hoverbear) [inline src/sources/socket/mod.rs:37]: Are these only used for tests? Might be able to put them behind a `#[cfg(test)]`

File: src/tcp.rs [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    #[serde(with = "duration")]
+    pub time: Option<Duration>,
+    #[serde(default)]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(with = "duration")]
+    pub interval: Option<Duration>,
+    pub retries: Option<u32>,
+}
+
+mod duration {
```
Comment (Hoverbear) [inline src/tcp.rs:31]: This release suggests this is not needed?https://github.com/serde-rs/serde/releases/tag/v0.8.1

Comment (jszwedko) [reply src/tcp.rs:31]: Now noticing also that these config options are lacking the `_secs` suffix we usually have for duration parameters.

File: src/sources/statsd/mod.rs [hunk 7]
```diff
@@ -39,6 +40,7 @@ pub struct UdpConfig {
 #[derive(Deserialize, Serialize, Debug, Clone)]
 struct TcpConfig {
     address: SocketListenAddr,
+    pub keepalive: Option<TcpKeepaliveConfig>,
```
Comment (bruceg) [inline src/sources/statsd/mod.rs:43]: AFAICT this does not need to be `pub` at all.

File: src/tcp.rs [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+/// has no effect on a TCP stream in vector. These options are parsed into the options config
+/// nevertheless for future support.
+///
+/// Implementing these options would require upgrading the Tokio runtime accordingly first.
+#[derive(Clone, Copy, Debug, Deserialize, PartialEq, Serialize)]
+#[serde(deny_unknown_fields)]
+pub struct TcpKeepaliveConfig {
+    #[serde(default)]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(with = "duration")]
```
Comment (bruceg) [inline src/tcp.rs:22]: I think these can be joined:
```suggestion
    #[serde(default, skip_serializing_if = "Option::is_none", with = "duration")]
```

File: src/tcp.rs (hunk 6 above)
Comment (pablosichert) [reply src/tcp.rs:31]: `Duration` is [serialized/deserialized](https://github.com/serde-rs/serde/blob/2b504099e4e480625e74d3599f5ddab2f42c63de/serde/src/de/impls.rs#L1857-L1860) as:

```rust
struct Duration {
    secs: u64,
    nanos: u32,
 }
```

Considering that second precision is more than enough for the timeout, I thought reading/writing the value as a plain `u64` makes more sense in this case. I chose to deserialize as `Duration` to not lose the semantic information of it being a time value.

> Now noticing also that these config options are lacking the _secs suffix we usually have for duration parameters.

Thanks for the hint, I guess it makes sense to do the same if this is the pattern: https://github.com/timberio/vector/blob/95f767d02f92cb6e320b8dc51371dd02959f9499/src/sinks/util/service.rs#L166-L171

File: src/sources/socket/mod.rs (hunk 5 above)
Comment (pablosichert) [reply src/sources/socket/mod.rs:37]: I liked your suggestion and tried, but it doesn't work for two reasons:
1) it's used in benches
2) it's used in integration tests which build the module without the test config

File: src/tcp.rs (hunk 4 above)
Comment (pablosichert) [reply src/tcp.rs]: Done in 59c042f.

File: src/tcp.rs (hunk 6 above)
Comment (pablosichert) [reply src/tcp.rs]: Done in b0e5caf.

File: src/sinks/util/tcp.rs [hunk 9]
```diff
@@ -49,12 +50,21 @@ enum TcpError {
 #[serde(deny_unknown_fields)]
 pub struct TcpSinkConfig {
     pub address: String,
+    pub keepalive: Option<TcpKeepaliveConfig>,
     pub tls: Option<TlsConfig>,
```
[the remark this answers is not in the document]
Comment (pablosichert) [reply src/sinks/util/tcp.rs]: Done in 5f46c35.

File: src/tcp.rs (hunk 8 above)
Comment (pablosichert) [reply src/tcp.rs]: Done in 9ea5c7e, but obsolete due to b0e5caf. Thanks for the hint anyway!

File: src/sinks/socket.rs [hunk 10]
```diff
@@ -46,16 +47,21 @@ impl SocketSinkConfig {
...
+        keepalive: Option<TcpKeepaliveConfig>,
         tls: Option<TlsConfig>,
     ) -> Self {
         SocketSinkConfig {
-            mode: Mode::Tcp(TcpSinkConfig { address, tls }),
+            mode: Mode::Tcp(TcpSinkConfig {
+                address,
+                keepalive,
+                tls,
+            }),
```
[the remark this answers is not in the document]
Comment (pablosichert) [reply src/sinks/socket.rs]: Done in 5f46c35.

**Comment by pablosichert** [comment]
Thank a lot for your reviews and comments!

File: src/tls/incoming.rs [hunk 11]
```diff
@@ -158,6 +160,26 @@ impl MaybeTlsIncomingStream<TcpStream> {
         Ok(())
     }
 
+    // Required for `make check-component-features`
+    #[allow(dead_code)]
+    pub(crate) fn set_keepalive(
```
Comment (bruceg) [inline src/tls/incoming.rs:165]: Do you know which features require this function? We would prefer to feature gate it (ie `#[cfg(all(feature = "X", feature = "Y"))]` rather than use this dead code allowance.

**Review by bruceg — COMMENTED** [review]
This is looking better, just one question about a dead code flag.

File: src/tls/incoming.rs (hunk 11 above)
Comment (pablosichert) [reply src/tls/incoming.rs:165]: Agree, I don't like this part here, at all.

My first approach was introducing a `tcp` flag which `tls` depends on, and put the `set_keepalive` behind the `tcp` flag. Unfortunately, that didn't solve the problem, as it still showed as unused code. I feel making this work as intended would require to refactor larger parts of the `tcp` util. Maybe we make a separate ticket for that?

**Review by Hoverbear — APPROVED** [review]
Nice! Thanks. :)

**Comment by jamtur01** [comment]
@pablosichert Can you rebaase and fix the merge conflict when you get a chance? Thanks.

File: src/tls/incoming.rs (hunk 11 above)
Comment (jamtur01) [reply src/tls/incoming.rs:165]: @bruceg Refactor or ticket? I lean towards ticket.

Comment (pablosichert) [reply src/tls/incoming.rs:165]: Looking into that now.

Comment (pablosichert) [reply src/tls/incoming.rs]: Done in 8e21651. Was less invasive than I anticipated, now after the flags have been cleaned up on master.

File: src/tls/incoming.rs [hunk 12]
```diff
@@ -158,6 +162,25 @@ impl MaybeTlsIncomingStream<TcpStream> {
         Ok(())
     }
 
+    #[cfg(feature = "sources-utils-tcp-keepalive")]
+    pub(crate) fn set_keepalive(
+        &mut self,
+        keepalive: Option<TcpKeepaliveConfig>,
+    ) -> io::Result<()> {
```
Comment (bruceg) [inline src/tls/incoming.rs:169]: Since the underlying function takes the value directly and not an `Option`, would it make sense to have this take `TcpKeepaliveConfig` directly and move the `if let Some` into the callers? There appears to only be one caller, so it shouldn't make much difference.

File: src/tls/mod.rs [hunk 13]
```diff
@@ -124,6 +125,19 @@ impl MaybeTlsStream<TcpStream> {
             Self::Tls(tls) => tls.get_ref().peer_addr(),
         }
     }
+
+    pub fn set_keepalive(&mut self, keepalive: Option<TcpKeepaliveConfig>) -> std::io::Result<()> {
```
Comment (bruceg) [inline src/tls/mod.rs:129]: Ditto here.

**Review by bruceg — APPROVED** [review]
Just one non-blocking question that didn't occur to me before.

File: src/tls/incoming.rs (hunk 12 above)
Comment (pablosichert) [reply src/tls/incoming.rs:169]: Oh, that's a good point. I was fixating on https://docs.rs/tokio/0.2.23/tokio/net/struct.TcpStream.html#method.set_keepalive taking an `Option`, but it only relates to the `time`, which is also an `Option` in our struct.

Will address.

**Comment by jamtur01** [comment]
Awesome! Thanks @pablosichert
